### ISSUE TITLE
1m tcp syn queue cookies and alarms

### ISSUE THREAD

**Issue by dwreski** [netdata/netdata#6998]
##### Question summary
Hi, this question has been asked numerous times before, but there is no clear answer.

I am continually receiving alerts similar to this on a moderately busy Fedora 29 webserver with 128GB of RAM on a gigabit link.

1m tcp syn queue cookies = 1 cookies the number of times the TCP SYN queue of the kernel was full and sent SYN cookies, during the last minute

##### OS / Environment
uname -r
5.2.11-100.fc29.x86_64

rpm -q netdata
netdata-1.17.1-1.fc29.x86_64

netstat -s | grep -i listen
    15 SYNs to LISTEN sockets dropped

nstat -a | grep -i listen
TcpExtListenDrops               15                 0.0

##### Expected results
Attached is my sysctl.conf. Perhaps there is a kernel parameter I should be adjusting to fix this? This is also happening on multiple systems with similar configurations.

I've read #4281, #3234 and others and see no final solution. I'm not sure what other information I can provide to help troubleshoot this.

[sysctl.conf.log](https://github.com/netdata/netdata/files/3693622/sysctl.conf.log)

**Comment by scambra** [comment]
I have this alarm many times every day. No full drop or overflow, but cookies counter is really high:

nstat -za|grep -i -e tcpreq -e listen
TcpExtListenOverflows           0                  0.0
TcpExtListenDrops               35                 0.0
TcpExtTCPReqQFullDoCookies      43807              0.0
TcpExtTCPReqQFullDrop           0                  0.0
TcpExtTCPFastOpenListenOverflow 0                  0.0

**Comment by ilyam8** [comment]
The alarm is

https://github.com/netdata/netdata/blob/7b3f17d32425a2671cb733c4053b49cf6aa96b1c/health/health.d/tcp_listen.conf#L70-L82

___

[If syn cookies is enabled](https://blog.cloudflare.com/syn-packet-handling-in-the-wild/) (which is default):

> When a SYN cookie is being sent out:
 > - TcpExtTCPReqQFullDoCookies / LINUX_MIB_TCPREQQFULLDOCOOKIES is incremnted.
 > - TcpExtSyncookiesSent / LINUX_MIB_SYNCOOKIESSENT is incremented.
 > - Linux used to increment TcpExtListenDrops [but doesn't from kernel 4.7.](https://github.com/torvalds/linux/commit/9caad864151e525929d323de96cad382da49c3b2)


>  the number of times the TCP SYN queue of the kernel was full and sent SYN cookies

From what i see this is wrong. There is no `TCP SYN queue of the kernel was full` condition.

**Comment by ilyam8** [comment]
@literal i see [you suggested](https://github.com/netdata/netdata/issues/3234#issuecomment-423935842)

> And there is yet another related counter in /proc/net/netstat that I would suggest monitoring and adding an alarm for: TcpExtTCPReqQFullDoCookies?

I think [current alarms are wrong](https://github.com/netdata/netdata/blob/7b3f17d32425a2671cb733c4053b49cf6aa96b1c/health/health.d/tcp_listen.conf#L57-L82). Probably we didnt take into account some condition. What do you think?

**Comment by ilyam8** [comment]
ok, i do a search in the linux source code, probably it is related

probable syn flood conditions

https://github.com/torvalds/linux/blob/531e93d11470aa2e14e6a3febef50d9bc7bab7a1/net/ipv4/tcp_input.c#L6492-L6496

**Comment by Ferroin** [comment]
@ilyam8 SYN cookies are only sent if the SYN queue is full (or functionally full).  The idea is to be able to treat the queue as much larger than it actually is by encoding the data that would go into the SYN queue entry for the connection in the sequence number of the response. In such cases, if the response is legitimate, the kernel can then extract the info it needs from the return sequence number when the other side replies, and accept the connection even though the SYN queue is full.

So the alarm is technically correct, but understanding what exactly is going on typically requires a solid understanding of the network itself and multiple packet captures.

However, we should probably bump up the threshold a bit.  If it's only single isolated instances of the counter being incremented that we see, it's probably not an issue.  It's when it spikes really high (dozens or more in a single collection) or is consistently being incremented on a regular basis that we've got a likely issue.

**Comment by ilyam8** [comment]
I would say not a bit

```
 lookup: sum -60s unaligned absolute of TCPReqQFullDoCookies 
   warn: $this > 0 
   crit: $this > (($status == $CRITICAL) ? (0) : (60)) 
  delay: up 0 down 5m multiplier 1.5 max 1h 
```

I think we nees to change `warn` and `crit` conditions and `delay up`. `warn: $this > 0` doesnt look good to me at all.

Not sure about exact number tho 😄 

What would you suggest @Ferroin

**Comment by Ferroin** [comment]
* Keep the up delay low, this is something that could indicate an active attack on the system, so we want a fast response, probably make it something like 5-15 seconds.
* Switch to average instead of sum.  For this, we care about longer-term trends.
* Bump warn to 1. This functionally means that we'll ignore small spikes (if we are looking at 60s of data, we'll need a spike of more than 60 to trip the alarm at all).
* Drop crit to 5. That translates to 300 or more in a 60 second period, which is pretty conclusively an error.

We might be able to tune things a bit more, but I think this is probably enough.

**Comment by ilyam8** [comment]
@Ferroin are you willing to make a PR? 😄

**Comment by ilyam8** [comment]
should be fixed in #7250
